src/odemis/driver/pigcs.py

```diff
@@ -3070,16 +3101,57 @@ def moveAbs(self, pos, update=False):
         return f
     moveAbs.__doc__ = model.Actuator.moveAbs.__doc__
 
-    # TODO reference(self, axes)
-#     @isasync
-#     def reference(self, axes):
-#         if not axes:
-#             return model.InstantaneousFuture()
-#         self._checkReference(axes)
-#
-#         f = self._executor.submit(self._doReference, axes)
-#         return f
-#     reference.__doc__ = model.Actuator.reference.__doc__
+    @isasync
+    def reference(self, axes):
+        if not axes:
+            return model.InstantaneousFuture()
+        self._checkReference(axes)
+
+        f = self._createFuture(axes, False)
+        f = self._executor.submitf(f, self._doReference, f, axes)
+
+        return f
+    reference.__doc__ = model.Actuator.reference.__doc__
+
+    def _doReference(self, future, axes):
+        """
+        Actually runs the referencing code.
+        :param axes: (set of str) The axes that should be referenced.
+        :raises:
+            IOError: if referencing failed due to hardware
+            CancelledError: if was cancelled
+        """
+        # Reset reference so that if it fails, it states the axes are not
+        # referenced (anymore)
+        with future._moving_lock:
+            try:
+                # do the referencing for each axis sequentially
+                # (because each referencing is synchronous)
+                for a in axes:
+                    if future._must_stop.is_set():
+                        raise CancelledError()
+                    controller, channel = self._axis_to_cc[a]
+                    self.referenced._value[a] = False
+                    controller.startReferencing(channel)
+                    self._waitEndMove(future, (a,), time.time() + 100)  # block until it's over
+                    self.referenced._value[a] = True
+            except CancelledError:
+                # FIXME: if the referencing is stopped, the device refuses to
+                # move until referencing is run (and successful).
+                # => Need to put back the device into a mode where at least
+                # relative moves work.
+                logging.warning("Referencing cancelled, device will not move until another referencing")
```

Comment:
That looks like a left-over from copy-pasting, but I don't think this is correct for the PI controllers?